src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -216,45 +216,56 @@ internal sealed partial class VisualStudioProject
-            // that may end up touching the UI thread (As we can deadlock if GetTagsSynchronous waits on us).  Instead,
-            // we only check if the Task is completed.  Prior to that we will assume we are still loading.  Once this
-            // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
-            // fully loaded.
-            var isFullyLoadedTask = _workspaceStatusService?.IsFullyLoadedAsync(CancellationToken.None);
-            var isFullyLoaded = isFullyLoadedTask is { IsCompleted: true } && isFullyLoadedTask.GetAwaiter().GetResult();
+                field = newValue;
 
-            // We only log telemetry during solution open
-            if (logThrowAwayTelemetry && _telemetryService?.HasActiveSession == true && !isFullyLoaded)
-            {
-                TryReportCompilationThrownAway(_workspace.CurrentSolution.State, Id);
-            }
+                // Importantly, we do not await/wait on the fullyLoadedStateTask.  We do not want to ever be waiting on work
+                // that may end up touching the UI thread (As we can deadlock if GetTagsSynchronous waits on us).  Instead,
+                // we only check if the Task is completed.  Prior to that we will assume we are still loading.  Once this
+                // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
+                // fully loaded.
+                var isFullyLoadedTask = _workspaceStatusService?.IsFullyLoadedAsync(CancellationToken.None);
+                var isFullyLoaded = isFullyLoadedTask is { IsCompleted: true } && isFullyLoadedTask.GetAwaiter().GetResult();
 
-            if (_activeBatchScopes > 0)
-            {
-                _projectPropertyModificationsInBatch.Add(withNewValue);
-            }
-            else
-            {
-                _workspace.ApplyChangeToWorkspace(Id, withNewValue);
+                // We only log telemetry during solution open
+                if (logThrowAwayTelemetry && _telemetryService?.HasActiveSession == true && !isFullyLoaded)
+                {
+                    TryReportCompilationThrownAway(_workspace.CurrentSolution.State, Id);
+                }
+
+                if (_activeBatchScopes > 0)
+                {
+                    _projectPropertyModificationsInBatch.Add(solutionChanges => updateSolution(solutionChanges, oldValue));
+                }
+                else
+                {
+                    _workspace.ApplyBatchChangeToWorkspaceMaybeAsync(useAsync: false, s =>
+                    {
+                        var solutionChanges = new SolutionChangeAccumulator(s);
+                        updateSolution(solutionChanges, oldValue);
+                        return solutionChanges;
+                    }).AsTask().GetAwaiter().GetResult();
```

Comment:
is it desirable to be sync here?  as opposed to async, whcih you push up all the way to callers.

Reply:
Same comment as here: https://github.com/dotnet/roslyn/pull/59638/files/1268f8f6cc2a854311b427ac6a9701fd9742286f..bcc398b3ede4373f53e2584e1ac8fe421b6ec5ac#r810348018 Sorry should have left that unresolved for you to see the answer.

Basically, the sync path here is taken by the legacy project system, that that point we're better off synchronously acquiring the locks since that'll get priority. So other than switching to be a JTF.Run() higher level, we can't really make this properly sync. And I suspect that might actually be worse for perf given SemaphoreSlim's prioritization for synchronous waters.

Reply:
Somehow you broke GitHub, the current code here anyways is:

                    _workspace.ApplyBatchChangeToWorkspace(solutionChanges =>
                    {
                        updateSolution(solutionChanges, oldValue);
                    });

So the wart for the .GetResult() is gone.